File: parse_assets.R
library(tidyverse)
library(jsonlite)

data = read_tsv('./assets/htan-assets-bucket.tsv', col_names = c("bucket", "key", "modified")) %>%
    filter(
        grepl("synid\\/syn", key) & grepl("thumbnail.[png|jpg]|index.html$", key)
    ) %>%
    extract(
        key,
        into = c("synid", "version", "type"),
        regex = "synid\\/(syn\\d{8})\\/(.+)\\/(thumbnail|minerva)",
        remove = FALSE
    ) %>%
    mutate(
        s3_uri = paste0("s3://", bucket, "/", key),
        cloudfront_url = paste0("https://d3p249wtgzkn5u.cloudfront.net/", key)
    )

write_csv(data, './assets/htan-assets-tidy.csv')

latest <- data %>%
   group_by(synid, type) %>%
    filter(modified == max(modified))
    
write_csv(latest, "assets/htan-assets-latest.csv")


#inerva_thumbs <- read_csv('minerva-thumbs.csv') %>% rename(new = thumbnail)
#inerva_thumbs
#
#inerva_thumbs <- read_tsv('./assets/htan-assets-bucket.tsv', col_names = c("bucket", "key", "modified")) %>%
#   filter(
#       grepl("synid\\/syn", key) & grepl("_0_0.jpg$", key)
#   ) %>%
#   extract(
#       key,
#       into = c("synid", "version"),
#       regex = "synid\\/(syn\\d{8})\\/(.+)\\/",
#       remove = FALSE
#   ) %>%
#   mutate(
#       level = str_match(key, "(\\d+)(_0_0.jpg)$")[,2]
#   )
#
#inerva_thumbs

latest
#portal_images <- "https://github.com/ncihtan/htan-portal/raw/master/public/processed_syn_data.json.gz"
#
#temp <- tempfile()
#download.file(portal_images, temp)
#gzfile(temp, 'rt')
#portal_images <- jsonlite::read_json(temp)
#unlink(temp)
#
#portal_images <- portal_images$files %>% 
#enframe() %>% unnest_wider(value) %>% filter(Component == "ImagingLevel2")

latest %>%
    ungroup() %>%
    select(-s3_uri,-version, -modified,-bucket,-key) %>%
    pivot_wider(names_from = type, values_from = cloudfront_url) %>%
    jsonlite::write_json('assets/htan-imaging-assets.json', pretty = TRUE)

#latest %>%
#    ungroup() %>%
#    select(-s3_uri,-version, -modified,-bucket,-key) %>%
#    pivot_wider(names_from = type, values_from = cloudfront_url) %>%
#    left_join(minerva_thumbs, by = c('synid','minerva')) %>%
#    filter(!grepl("thirsty_mclean", thumbnail)) %>%
#    mutate(thumbnail = case_when(is.na(thumbnail)) ~ new, TRUE ~ thumbnail) %>%
#    select(-new)
#    jsonlite::write_json('assets/htan-imaging-assets-fixed.json', pretty = TRUE)

File: parse_assets.py
import pandas as pd
import json
import re

# Reading the TSV file
data = pd.read_csv('./assets/htan-assets-bucket.tsv', sep='\t', names=["bucket", "key", "modified"])

# Filtering rows
data = data[data['key'].str.contains('synid/syn') & data['key'].str.contains('thumbnail\.png|jpg|index\.html$')]

data = data[~data['key'].str.contains('nasty_ekeblad', na=False)]


# Extracting new columns
data[['synid', 'version', 'type']] = data['key'].str.extract(r'synid/(syn\d{8})/(.+?)/((?:thumbnail(?:\.png|\.jpg)|minerva/index\.html)?)', expand=True)

data = data[pd.notna(data['type'])]

data['type'] = data['type'].map({
    'minerva/index.html': 'minerva',
    'thumbnail.png': 'thumbnail',
    'thumbnail.jpg': 'thumbnail'
}, na_action='ignore')

# Adding new columns for URLs
data['s3_uri'] = 's3://' + data['bucket'] + '/' + data['key']
data['cloudfront_url'] = 'https://d3p249wtgzkn5u.cloudfront.net/' + data['key']


# Writing the processed data to a CSV file
data.to_csv('./assets/htan-assets-tidy.csv', index=False)

# Identifying the latest entries
latest = data.groupby(['synid', 'type']).apply(lambda x: x[x['modified'] == x['modified'].max()])

# Writing the latest entries to a CSV file
latest.to_csv('./assets/htan-assets-latest.csv', index=False)

# Additional transformations for JSON output
latest = latest.reset_index(drop=True)
latest = latest.drop(['s3_uri', 'version', 'modified', 'bucket', 'key'], axis=1)
latest_json = latest.pivot_table(index=['synid'], columns='type', values='cloudfront_url', aggfunc='first').reset_index()


# Writing to a JSON file
with open('assets/htan-imaging-assets.json', 'w', encoding='utf-8') as f:
    json.dump(latest_json.to_dict(orient='records'), f, ensure_ascii=False, indent=4)
